### PR TITLE
feature: Improved Recurrence Invitations Messages

### LINKED ISSUE #38686: [Bug]: Event update email unclear with recurrence exceptions
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Deleting an instance of a recurring event sends out an event update/invite without any useful information about what really changed.

The email says "Time: <time and date of very first occurence>"

### Steps to reproduce

1. Create an event with one attendee that recurs weekly for three times
2. Accept the invitation as the other user
3. Delete the second occurrence as the organizer

```ics
BEGIN:VCALENDAR
VERSION:2.0
CALSCALE:GREGORIAN
BEGIN:VEVENT
CREATED:20230607T094533Z
DTSTAMP:20230607T094848Z
LAST-MODIFIED:20230607T094848Z
SEQUENCE:4
UID:5e8a4ae6-39d7-4cb1-91e0-f04036be45d9
DTSTART;TZID=Europe/Vienna:20230607T133000
DTEND;TZID=Europe/Vienna:20230607T153000
STATUS:CONFIRMED
SUMMARY:Zoo party
ATTENDEE;CN=schiraffe@domain.tld;CUTYPE=INDIVIDUAL;PARTSTAT=ACCEPTED;ROL
 E=REQ-PARTICIPANT;SCHEDULE-STATUS=1.1:mailto:schiraffe@domain.tld
ORGANIZER;CN=Christoph Wurst:mailto:org@domain.tld
RRULE:FREQ=WEEKLY;COUNT=3;BYDAY=WE
LOCATION:At the zoo
DESCRIPTION:Bring your own food
EXDATE;TZID=Europe/Vienna:20230614T133000
BEGIN:VALARM
ACTION:DISPLAY
TRIGGER;RELATED=START:-PT1H
END:VALARM
END:VEVENT
END:VCALENDAR
```

### Expected behavior

The attendee receives an email that says something like "Time: <time and date of very first occurence>, recurring weekly except <new exception date>".

### Installation method

None

### Nextcloud Server version

master

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
n/a
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

https://github.com/nextcloud/calendar/issues/3919

### PR DESCRIPTION
**Pull request by SebastianKrupinski**
* Resolves: #38686

## Summary
Improved Recurrence Invitations Messages

### CODE REVIEW

File: apps/dav/lib/CalDAV/EventReaderRRule.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	public function precision(): string {
+		return $this->frequency;
+	}
+
+	public function interval(): int {
+		return $this->interval;
+	}
+
+	public function concludes(): \DateTime | null {
+		if (isset($this->until)) {
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/EventReaderRRule.php:39]:
```suggestion
		if ($this->until instanceof \DateTimeInterface) {
```

isset is totally fine because until is typed as \DateTimeInterface|null. 
I prefer the instanceof version for the better readability.

File: apps/dav/lib/CalDAV/EventReaderRRule.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+			return \DateTime::createFromImmutable($this->until);
+		} elseif ($this->count > 0) {
+			// temporarly store current reccurance date and counter
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/EventReaderRRule.php:42]:
```suggestion
			// temporarily store current recurrence date and counter
```

File: apps/dav/lib/CalDAV/EventReaderRRule.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			$currentReccuranceDate = $this->currentDate;
+			$currentCounter = $this->counter;
+			// iterate over occurances until last one (subtract 2 from count for start and end occurance)
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/EventReaderRRule.php:45]:
```suggestion
			// iterate over occurrences until last one (subtract 2 from count for start and end occurrence)
```

File: apps/dav/lib/CalDAV/EventReaderRRule.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			while ($this->counter <= ($this->count - 2)) {
+				$this->next();
+			}
+			// temporarly store last reccurance date
+			$lastReccuranceDate = $this->currentDate;
+			// restore current reccurance date and counter
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/EventReaderRRule.php:51]:
```suggestion
			// restore current recurrence date and counter
```

File: apps/dav/lib/CalDAV/EventReaderRRule.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			$this->currentDate = $currentReccuranceDate;
+			$this->counter = $currentCounter;
+			// return last recurrance date
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/EventReaderRRule.php:54]:
```suggestion
			// return last recurrence date
```

File: apps/dav/lib/CalDAV/EventReader.php [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+	 *
+	 * The $uid parameter is only required for the first method.
+	 *
+	 * @since 30.0.0
+	 *
+	 * @param VCalendar|VEvent|Array|String $input
+	 * @param string|null     				$uid
+	 * @param DateTimeZone|null    				$timeZone reference timezone for floating dates and times
+	 */
+	public function __construct(VCalendar|VEvent|array|string $input, ?string $uid = null, ?DateTimeZone $timeZone = null) {
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/EventReader.php:70]: Please split this to multiple static constructor functions and keep the constructor minimal. The `if`/`elseif` statements are hard to follow and maintain.

It will also be easier to deprecate some of the functions in the future.

E.g.
```
public static function fromVCalendar(/* [...] */): self {
}

public static function fromVEventArray(/* [...] */): self {
}

// etc.
```

File: apps/dav/lib/CalDAV/EventReaderRDate.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+	public function concludesOn(): \DateTime | null {
+		if (count($this->dates) > 0) {
+			return new \DateTime(
+				$this->dates[array_key_last($this->dates)],
+				$this->startDate->getTimezone()
+			);
+		} else {
+			return null;
+		}
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/EventReaderRDate.php:30]:
```suggestion
		
		return null;
```

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 8]
```diff
@@ -158,115 +167,295 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+	public function generateWhenStringRecurringDaily(EventReader $er): string {
+		
+		// initialize
+		$interval = (int) $er->recurringInterval();
+		$startTime = '';
+		$endTime = '';
+		$conclusion = '';
+		// time of the day
+		if (!$er->entireDay()) {
+			$startTime = $this->l10n->l('time', $er->startDateTime(), ['width' => 'short']) . (($er->startTimeZone() != $er->endTimeZone()) ? ' (' . $er->startTimeZone()->getName() . ')': '');
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:243]: Please refactor the ternary operator. This line is hardly readable.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 9]
```diff
@@ -158,115 +167,295 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+		// initialize
+		$interval = (int) $er->recurringInterval();
+		$startTime = '';
+		$endTime = '';
+		$conclusion = '';
+		// days of the week
+		$days = implode(', ', array_map(function ($value) { return $this->l10n->t($value); }, $er->recurringDaysOfWeekNamed()));
+		// time of the day
+		if (!$er->entireDay()) {
+			$startTime = $this->l10n->l('time', $er->startDateTime(), ['width' => 'short']) . (($er->startTimeZone() != $er->endTimeZone()) ? ' (' . $er->startTimeZone()->getName() . ')': '');
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:285]: Please refactor the ternary operator. This line is hardly readable.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 10]
```diff
@@ -158,115 +167,295 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+		// days of month
+		if ($er->recurringPattern() === 'R') {
+			$days = implode(', ', array_map(function ($value) { return $this->l10n->t($value); }, $er->recurringRelativePositionNamed())) . ' ' .
+					implode(', ', array_map(function ($value) { return $this->l10n->t($value); }, $er->recurringDaysOfWeekNamed()));
+		} else {
+			$days = implode(', ', $er->recurringDaysOfMonth());
+		}
+		// time of the day
+		if (!$er->entireDay()) {
+			$startTime = $this->l10n->l('time', $er->startDateTime(), ['width' => 'short']) . (($er->startTimeZone() != $er->endTimeZone()) ? ' (' . $er->startTimeZone()->getName() . ')': '');
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:332]: Please refactor the ternary operator. This line is hardly readable.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 11]
```diff
@@ -158,115 +167,295 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+	 */
+	public function generateWhenStringRecurringWeekly(EventReader $er): string {
+		
+		// initialize
+		$interval = (int) $er->recurringInterval();
+		$startTime = '';
+		$endTime = '';
+		$conclusion = '';
+		// days of the week
+		$days = implode(', ', array_map(function ($value) { return $this->l10n->t($value); }, $er->recurringDaysOfWeekNamed()));
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:282]: I'm not sure if `$this->l10n->t($value)` will work. AFAIK, the strings inside `t()` have to be literals and variables can't be used directly.

Did you test it using a different language setting?

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 12]
```diff
@@ -145,7 +154,7 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
 		}
 
 		if(!empty($oldVEvent)) {
-			$oldMeetingWhen = $this->generateWhenString($oldVEvent);
+			$oldMeetingWhen = $eventReaderPrevious->recurs() ? $this->generateWhenStringRecurring($eventReaderPrevious) : $this->generateWhenStringSingular($eventReaderPrevious);
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:157]: It makes sense to keep the `generateWhenString()` method and refactor it a bit. The result of this ternary operator is used multiple times and its code is currently duplicated.

```php
private function generateWhenString(EventReader $reader): string {
	if ($eventReaderPrevious->recurs()) {
		return $this->generateWhenStringRecurring($eventReaderPrevious);
	}

 	return $this->generateWhenStringSingular($eventReaderPrevious);
}
```

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 13]
```diff
@@ -130,9 +135,13 @@ private function linkify(?string $url): ?string {
...
 	 */
 	public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
+
+		// construct event reader
+		$eventReaderCurrent = new EventReader($vEvent);
+		$eventReaderPrevious = !empty($oldVEvent) ? new EventReader($oldVEvent) : null;
 		$defaultVal = '';
 		$data = [];
-		$data['meeting_when'] = $this->generateWhenString($vEvent);
+		$data['meeting_when'] = $eventReaderCurrent->recurs() ? $this->generateWhenStringRecurring($eventReaderCurrent) : $this->generateWhenStringSingular($eventReaderCurrent);
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:144]: See above regarding `generateWhenString()`.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 14]
```diff
@@ -158,115 +167,295 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
 	 * @return array
 	 */
 	public function buildCancelledBodyData(VEvent $vEvent): array {
+		// construct event reader
+		$eventReaderCurrent = new EventReader($vEvent);
 		$defaultVal = '';
 		$strikethrough = "<span style='text-decoration: line-through'>%s</span>";
 
-		$newMeetingWhen = $this->generateWhenString($vEvent);
+		$newMeetingWhen = $eventReaderCurrent->recurs() ? $this->generateWhenStringRecurring($eventReaderCurrent) : $this->generateWhenStringSingular($eventReaderCurrent);
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:457]: See above regarding `generateWhenString()`.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 15]
```diff
@@ -158,115 +167,295 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+		return match ([($interval > 1), !empty($startTime), !empty($conclusion)]) {
+			[false, false, false] => $this->l10n->t('Every Week on %1$s for the entire day', [$days]),
+			[false, false, true] => $this->l10n->t('Every Week on %1$s for the entire day until %2$s', [$days, $conclusion]),
+			[false, true, false] => $this->l10n->t('Every Week on %1$s between %2$s - %3$s', [$days, $startTime, $endTime]),
+			[false, true, true] => $this->l10n->t('Every Week on %1$s between %2$s - %3$s until %4$s', [$days, $startTime, $endTime, $conclusion]),
+			[true, false, false] => $this->l10n->t('Every %1$d Weeks on %2$s for the entire day', [$interval, $days]),
+			[true, false, true] => $this->l10n->t('Every %1$d Weeks on %2$s for the entire day until %3$s', [$interval, $days, $conclusion]),
+			[true, true, false] => $this->l10n->t('Every %1$d Weeks on %2$s between %3$s - %4$s', [$interval, $days, $startTime, $endTime]),
+			[true, true, true] => $this->l10n->t('Every %1$d Weeks on %2$s between %3$s - %4$s until %5$s', [$interval, $days, $startTime, $endTime, $conclusion]),
+			default => $this->l10n->t('Could not generate event recurrence statement')
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:302]: I think it would be better to throw here because all 8 cases are covered anyway.

E.g. `throw new \Exception('Unreachable match case');`

The same also applies to the other similar methods in this class.

**Review by st3iny — CHANGES_REQUESTED** [review]

**Review by st3iny — COMMENTED** [review]
In general there are a lot of ternary operators. I'm not a big fan because they sacrifice readability for brevity.

I tagged the ones that are a bit problematic from my point of view. I'd also prefer some of the others to be refactored to regular if statements but have no hard opinions.

E.g. there are a lot of `return cond ? a : b` constructs.

---

It would also be nice to have some exemplary screenshots of the new messages or concrete steps for testing.

File: apps/dav/lib/CalDAV/EventReader.php (hunk 6 above)
Comment (kesselb) [reply apps/dav/lib/CalDAV/EventReader.php:70]: Good suggestion. 
Do we need all those ways to initialize the reader?

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 11 above)
Comment (ChristophWurst) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:282]: They have to be literals, yes. The l10n extraction tool can only look at the code statically. Variables can't be processed

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 15 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:302]: @st3iny so I though of doing this but the trade off is that throwing an error would cause the event NOT to be saved. Since the email sending is done before we save the event to the database, I didn't want a missed match case to cause data loss.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 12 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:157]: @st3iny you're right make sense to handle the if in a sub routine.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 8 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:243]: How is this for a readability?

```
$startTime = $this->l10n->l('time', $er->startDateTime(), ['width' => 'short']);
$startTime .= $er->startTimeZone() != $er->endTimeZone() ? ' (' . $er->startTimeZone()->getName() . ')' : '';
```

**Comment by SebastianKrupinski** [comment]
> It would also be nice to have some exemplary screenshots of the new messages or concrete steps for testing.

For testing just create a new calendar event with at least one attendee that has a valid email address.

Here are some screenshots of the new messages...

Singleton Partial Day Event

Screenshot 2024-06-25 174246

Singleton Complete Day Event

Screenshot 2024-06-25 175756

Recurring Daily Event

Screenshot 2024-06-25 174523

Recurring Weekly Event

Screenshot 2024-06-25 174959

Recurring Bi-Weekly Event

Screenshot 2024-06-25 175235

Recurring Month Event with Absolute Date

Screenshot 2024-06-25 175425

Recurring Monthly Event with Relative Date

Screenshot 2024-06-25 175615

File: apps/dav/lib/CalDAV/EventReader.php (hunk 6 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/EventReader.php:70]: @kesselb 
So I modeled this constructor after the original EventIterator from SaberDAV...

https://github.com/sabre-io/vobject/blob/1b388aab5cdbcb75a09be25f2e9119fc4463ae59/lib/Recur/EventIterator.php#L92

They have 3 different input formats, VCalendar, VEvent and an array of VEvents. I added the string so that the class can be more flexible and use full in the future.

File: apps/dav/lib/CalDAV/EventReaderRDate.php (hunk 7 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/EventReaderRDate.php:30]: > 

This would be a redundant return of the same value... solved it by removing the else and brakets

**Review by ChristophWurst — APPROVED** [review]

File: apps/dav/lib/CalDAV/EventReader.php [hunk 16]
```diff
@@ -0,0 +1,772 @@
...
+			// return the highest date
+			return (($rdate > $rrule) ? $rdate : $rrule);
+		} elseif (isset($rrule)) {
+			return $rrule;
+		} elseif (isset($rdate)) {
+			return $rdate;
+		}
+
+		return null;
+
```
Comment (miaulalala) [inline apps/dav/lib/CalDAV/EventReader.php:426]:
```suggestion

		if (is_null($this->rruleIterator) && is_null($this->rdateIterator)) {
			return null;
		}
		
		// retrieve rrule conclusion date
		if (isset($this->rruleIterator)) {
			$rrule = $this->rruleIterator->concludes();
			// evaluate if rrule conclusion is null
			// if this is null that means the recurrence is infinte
			if ($rrule === null) {
				return null;
			}
		}
		
		// retrieve rdate conclusion date
		if (isset($this->rdateIterator)) {
			$rdate = $this->rdateIterator->concludes();
		}
		
		// evaluate if both rrule and rdate have date
		if (isset($rdate) && isset($rrule)) {
			// return the highest date
			return (($rdate > $rrule) ? $rdate : $rrule);
		} 
		
		if (isset($rrule)) {
			return $rrule;
		} 
			
		return $rdate;

```

File: apps/dav/lib/CalDAV/EventReader.php [hunk 17]
```diff
@@ -0,0 +1,772 @@
...
+		// evaluate if relative position is set
+		if (is_array($positions)) {
+			// convert numberic relative position to relative label
+			foreach ($positions as $key => $value) {
+				$positions[$key] = $this->relativePositionNamesMap[$value];
+			}
+			return $positions;
+		}
+		// return empty array if evaluation failed
+		return [];
```
Comment (miaulalala) [inline apps/dav/lib/CalDAV/EventReader.php:631]: same early return as per above

File: apps/dav/lib/CalDAV/EventReader.php [hunk 18]
```diff
@@ -0,0 +1,772 @@
...
+	/**
+	 * event recurrance date
+	 *
+	 * returns date of currently selected recurrance
+	 *
+	 * @since 30.0.0
+	 *
+	 * @return DateTime
+	 */
+	public function recurrenceDate(): DateTime | null {
```
Comment (miaulalala) [inline apps/dav/lib/CalDAV/EventReader.php:643]:
```suggestion
	public function recurrenceDate(): ?DateTime {
```

File: apps/dav/lib/CalDAV/EventReader.php [hunk 19]
```diff
@@ -0,0 +1,772 @@
...
+		if ($this->recurrenceCurrentDate !== null) {
+			return DateTime::createFromImmutable($this->recurrenceCurrentDate);
+		} else {
+			return null;
+		}
```
Comment (miaulalala) [inline apps/dav/lib/CalDAV/EventReader.php:648]:
```suggestion
		return null;
```

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 20]
```diff
@@ -668,10 +879,104 @@ public function isRoomOrResource(Property $attendee): bool {
...
+			$scale = ($dateInterval->d > 1) ? 'days' : 'day';
+		} elseif ($dateInterval->h > 0) {
+			$interval = $dateInterval->h;
+			$scale = ($dateInterval->h > 1) ? 'hours' : 'hour';
+		} else {
+			$interval = $dateInterval->i;
+			$scale = 'minutes';
+		}
+
+		return [$interval, $scale];
```
Comment (miaulalala) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:915]:
```suggestion
		if ($dateInterval->y > 0) {
			$interval = $dateInterval->y;
			$scale = ($dateInterval->y > 1) ? 'years' : 'year';
			return [$interval, $scale];
		} 
		if ($dateInterval->m > 0) {
			$interval = $dateInterval->m;
			$scale = ($dateInterval->m > 1) ? 'months' : 'month';
			return [$interval, $scale];
		} 
		if ($dateInterval->d >= 7) {
			$interval = (int)($dateInterval->d / 7);
			$scale = ((int)($dateInterval->d / 7) > 1) ? 'weeks' : 'week';
			return [$interval, $scale];
		} 
		if ($dateInterval->d > 0) {
			$interval = $dateInterval->d;
			$scale = ($dateInterval->d > 1) ? 'days' : 'day';
			return [$interval, $scale];
		} 
		if ($dateInterval->h > 0) {
			$interval = $dateInterval->h;
			$scale = ($dateInterval->h > 1) ? 'hours' : 'hour';
			return [$interval, $scale];
		}
		
		$interval = $dateInterval->i;
		$scale = 'minutes';
		return [$interval, $scale];
```

**Review by miaulalala — CHANGES_REQUESTED** [review]

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 21]
```diff
@@ -145,128 +154,326 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+	 */
+	public function generateWhenStringSingular(EventReader $er): string {
+		// calculate time differnce from now to start of event
+		$occuring = $this->minimizeInterval($this->timeFactory->getDateTime()->diff($er->recurrenceDate()));
+		// extract start date
+		$startDate = $this->l10n->l('date', $er->startDateTime(), ['width' => 'full']);
+		// time of the day
+		if (!$er->entireDay()) {
+			$startTime = $this->l10n->l('time', $er->startDateTime(), ['width' => 'short']);
+			$startTime .= $er->startTimeZone() != $er->endTimeZone() ? ' (' . $er->startTimeZone()->getName() . ')' : '';
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:208]: This check is needed when an event started today with timezone A and ends tomorrow with timezone B, like the summer time shift?

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 22]
```diff
@@ -145,128 +154,326 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+			$endTime = $this->l10n->l('time', $er->endDateTime(), ['width' => 'short']) . ' (' . $er->endTimeZone()->getName() . ')';
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:209]: We don't need the check for the timezone here?

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 23]
```diff
@@ -145,128 +154,326 @@ public function buildBodyData(VEvent $vEvent, ?VEvent $oldVEvent): array {
...
+		}
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/Schedule/IMipService.php:210]: I think it would help a bit to move this block to an own function to reduce duplicate code. 

```php
	public function generateStartAndEndTime(EventReader $eventReader): array
	{
		if ($eventReader->entireDay()) {
			return ['', ''];
		}

		$startTime = $this->l10n->l('time', $eventReader->startDateTime(), ['width' => 'short']);
		$endTime = $this->l10n->l('time', $eventReader->endDateTime(), ['width' => 'short']);

		if ($eventReader->startTimeZone() !== $eventReader->endTimeZone()) {
			$startTime .= ' (' . $eventReader->startTimeZone()->getName() . ')';
			$endTime .= ' (' . $eventReader->endTimeZone()->getName() . ')';
		}

		return [$startTime, $endTime];
	}
```

And call it like `[$startTime, $endTime] = $this->generateStartAndEndTime($er);`

File: apps/dav/lib/CalDAV/EventReader.php [hunk 24]
```diff
@@ -0,0 +1,772 @@
...
+		
+		// construct count place holder
+		$count = 0;
+		// retrieve and add RRULE iterations count
+		$count += isset($this->rruleIterator) ? (int) $this->rruleIterator->concludesAfter() : 0;
+		// retrieve and add RDATE iterations count
+		$count += isset($this->rdateIterator) ? (int) $this->rdateIterator->concludesAfter() : 0;
+		// return count
+		return !empty($count) ? $count : null;
+
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/EventReader.php:385]: This would break the logic of the function and even if I fix the logic I would be double testing test variables for null.

But thank you for the suggestion of the NullSafe operator this will be very useful in other parts of the code.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 22 above)
Comment (kesselb) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:209]: Ah, I guess the idea is to show only one timezone if they are the same.

Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:209]: yup

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 21 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:208]: This is check is need to produce a different when line depending on weather an event is a full day or not

File: apps/dav/lib/CalDAV/Schedule/IMipService.php (hunk 23 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:210]: I would leave this as is, that way if design wants custom formatting for singleton events to differ from recurring event then it can be adjusted per case.

**Review by AndyScherzinger — APPROVED** [review]

**Review by st3iny — APPROVED** [review]
LGTM! Extensive unit tests are great.

**Review by miaulalala — APPROVED** [review]
